### PR TITLE
feat(l1): implement forkchoiceupdatedv2, newpayloadv2 and getpayloadv2

### LINKED ISSUE #892: Implement ForkChoiceUpdatedV2
This needs to be in place to allow some ForkChoiceUpdatedV3 hive tests.

### PR DESCRIPTION
**Pull request by ElFantasma**
**Motivation**

We need to support post merge `V1`, `V2` and `V3` for `GetPayload` `NewPayload` nd `ForkChoiceUpdated` functions

**Description**

Implemented `V2` for those functions

Closes #892, #1184, #1186

Makes to pass about 20 new hive tests

### CODE REVIEW

File: crates/networking/rpc/types/payload.rs [hunk 1]
```diff
@@ -35,10 +35,19 @@ pub struct ExecutionPayloadV3 {
...
+    #[serde(
+        skip_serializing_if = "Option::is_none",
+        with = "serde_utils::u64::hex_str_opt",
+        default = "Option::default"
+    )]
+    pub blob_gas_used: Option<u64>,
+    #[serde(
+        skip_serializing_if = "Option::is_none",
+        with = "serde_utils::u64::hex_str_opt",
+        default = "Option::default"
```
Comment (fmoletta) [inline crates/networking/rpc/types/payload.rs:48]: I think we can just use `default` here, we don't need to explicitly set the default value

File: crates/networking/rpc/engine/fork_choice.rs [hunk 2]
```diff
@@ -4,72 +4,74 @@ use ethrex_blockchain::{
...
+    fn try_from(fork_choice_updated: &dyn ForkChoiceUpdatedImpl) -> Result<RpcRequest, String> {
+        let request = fork_choice_updated.request();
+        match &request.payload_attributes {
+            Ok(attrs) => Ok(RpcRequest {
+                method: fork_choice_updated.method(),
+                params: Some(vec![
+                    serde_json::json!(request.fork_choice_state),
+                    serde_json::json!(attrs),
+                ]),
+                ..Default::default()
```
Comment (fkrause98) [inline crates/networking/rpc/engine/fork_choice.rs:51]: Nit: I'd prefer if we used some explicit fields here.

**Review by fkrause98 — APPROVED** [review]
LGTM!

File: crates/networking/rpc/engine/fork_choice.rs (hunk 2 above)
Comment (ElFantasma) [reply crates/networking/rpc/engine/fork_choice.rs:51]: I kind of prefer that too... but it's like [repeating arbitrary code](https://github.com/lambdaclass/ethrex/blob/9a2f8e0f86f1544aea25a6af66b34a3725b8a0c0/crates/networking/rpc/utils.rs#L190):

```rust
            id: RpcRequestId::Number(1),
            jsonrpc: "2.0".to_string(),
```

**Review by rodrigo-o — APPROVED** [review]
LGTM!

File: crates/networking/rpc/engine/fork_choice.rs [hunk 3]
```diff
@@ -157,3 +150,138 @@ impl RpcHandler for ForkChoiceUpdatedV3 {
...
+    type Error = String;
+
+    fn try_from(val: ForkChoiceUpdatedV2) -> Result<Self, Self::Error> {
+        ForkChoiceUpdated::try_from(&val)
+    }
+}
+
+impl RpcHandler for ForkChoiceUpdatedV2 {
+    fn parse(params: &Option<Vec<Value>>) -> Result<Self, RpcErr> {
+        Ok(Self(ForkChoiceUpdated::parse(params)?))
```
Comment (mpaulucci) [inline crates/networking/rpc/engine/fork_choice.rs:222]: where do you validate that params is in fact v2?

File: crates/networking/rpc/engine/fork_choice.rs [hunk 4]
```diff
@@ -112,34 +114,25 @@ impl RpcHandler for ForkChoiceUpdatedV3 {
...
-                        "forkChoiceV3 used to build pre-Cancun payload".to_string(),
-                    ));
-                }
-                if attributes.timestamp <= head_block.timestamp {
-                    return Err(RpcErr::InvalidPayloadAttributes(
-                        "invalid timestamp".to_string(),
-                    ));
-                }
+                let chain_config = storage.get_chain_config()?;
+                fork_choice_updated.validate(attributes, chain_config, head_block)?;
```
Comment (mpaulucci) [inline crates/networking/rpc/engine/fork_choice.rs:127]: So this is the only thing that is different between the FCU calls? What if there is any actions besides validating?

File: crates/networking/rpc/engine/fork_choice.rs (hunk 3 above)
Comment (ElFantasma) [reply crates/networking/rpc/engine/fork_choice.rs:222]: The only difference is the parent_beacon_block_root that shouldn't be present in V2. And it is checked in v2 `validate()` [here](https://github.com/lambdaclass/ethrex/pull/1297/files#diff-4af4eb82669a3403cab38afef8fb3b766f9ded917858c00d0f9c905fde5bcb53R188)

File: crates/networking/rpc/engine/fork_choice.rs (hunk 4 above)
Comment (ElFantasma) [reply crates/networking/rpc/engine/fork_choice.rs:127]: For this call the only difference are the validations and the `version` field in the `PayloadArgs` [here](https://github.com/lambdaclass/ethrex/pull/1297/files#diff-4af4eb82669a3403cab38afef8fb3b766f9ded917858c00d0f9c905fde5bcb53R135). If some other action or validation is required we can extend the `ForkChoiceUpdatedImpl` trait to handle it in each version that implements it

**Comment by ElFantasma** [comment]
Included in https://github.com/lambdaclass/ethrex/pull/1411

**Pull request closed without merging**